### PR TITLE
Add output_layer argument to M3GNet forward method to return intermediate layer outputs

### PR DESCRIPTION
**Pull request by JiQi535**
## Summary
In this PR, the `forward` method of M3GNet is modified so that intermediate outputs in "embedding", "gc_1", "gc_2", "gc_3", and "readout" layers can be extracted, while the "final" layer output stays as the default output. This allows the usage of M3GNet intermediate outputs as atom, bond, and structure features. Documentations and tests are provided.

Major changes:

1. Added the argument of `output_layer` into the `forward` method.
2. Added comprehensive tests for the `predict_structure` method.
3. Added the argument of `output_layer` into the `predict_structure` method.
4. Edited the variable names in the readout layer of M3GNet to be more precise. For example, the `node_vec` is renamed to `field_vec`, as this variable can either be node feature or edge feature instead of just node feature, depending on the argument `field`.

## Todos

Follow-up analysis for the effectiveness of atom features.

## Checklist

- [x] Google format doc strings added. Check with `ruff`.
- [x] Type annotations included. Check with `mypy`.
- [x] Tests added for new features/fixes.
- [ ] If applicable, new classes/functions/modules have [`duecredit`](https://github.com/duecredit/duecredit) `@due.dcite` decorators to reference relevant papers by DOI ([example](https://github.com/materialsproject/pymatgen/blob/91dbe6ee9ed01d781a9388bf147648e20c6d58e0/pymatgen/core/lattice.py#L1168-L1172))

### CODE REVIEW

**Comment by shyuep** [comment]
I think it is an extremely bad practice to put this in the predict_structure method. The predict_structure should return the expected target of the model. You should put it in a different method.

**Comment by kenko911** [comment]
Hi @JiQi535, thanks for your PR. I agree that we should have another class to extract features from structures instead of just putting in the function predict_structure, which is mainly for property prediction. Maybe we can call the new function "extract_features" and then we can also remove the unnecessary "final" option in the output_layer?

**Comment by JiQi535** [comment]
Thanks for the comments @shyuep @kenko911 . I have added a different method, `featurize_structure`, and the `predict_structure` method has been changed back to its original format to only return the expected target of the model. 

By default,  `featurize_structure` now returns the "readout" layer output and the "gc_3" layer output of intensive and extensive M3GNet models, respectively. Also, tests are updated, and a ValueError is raised if `output_layer` is not one of the allowed names in ["embedding", "gc_1", "gc_2", "gc_3", "readout", "final"].

**Comment by shyuep** [comment]
There is just way too much code and this is poor software design.
1. featurize_structure should simply return a dict of all intermediate and final layers. This way, there is no need to recompute every time you need one layer.
2. predict_structure should call featurize_structure in some way. Even better, the forward method actually stores the computed layers and you can access it immediately.

**Comment by JiQi535** [comment]
@shyuep The two suggestions have been realized in the updated code. My edits do not change the default output of `forward` method, so that other related codes, e.g., the `Potential` class, do not need change. This requires the addition of an argument in the forward method, i.e., `return_all_layer_output`. This implementation also does not increase the memory consumption of the `forward` method when `return_all_layer_output` is False as default.

File: src/matgl/models/_m3gnet.py [hunk 1]
```diff
@@ -251,35 +259,60 @@ def forward(
...
             graph_converter: Object that implements a get_graph_from_structure.
+            output_layers: List of names for the layer of GNN as output. Choose from "bond_expansion", "embedding",
+                "three_body_basis", "gc_1", "gc_2", "gc_3", "readout", and "final". By default, all M3GNet layer
+                outputs are returned.
 
         Returns:
-            output (torch.tensor): output property
+            output (dict): M3GNet intermediate and final layer outputs for a structure.
         """
+        allowed_output_layers = [
```
Comment (shyuep) [inline src/matgl/models/_m3gnet.py:301]: This cannot be hard coded. So if you have more than three GCLs or less than 3 GCLs, this code fails. Think about what kind of code you are creating. Are we saying M3GNet ALWAYS have 3 GCLs?

File: src/matgl/models/_m3gnet.py [hunk 2]
```diff
@@ -215,16 +215,19 @@ def forward(
         g: dgl.DGLGraph,
         state_attr: torch.Tensor | None = None,
         l_g: dgl.DGLGraph | None = None,
+        return_all_layer_output: bool = False,
```
Comment (shyuep) [inline src/matgl/models/_m3gnet.py:218]: Why is this needed? There is no compute cost to returning all layers right? Why not return all and then use the relevant layer?

File: src/matgl/models/_m3gnet.py (hunk 1 above)
Comment (JiQi535) [reply src/matgl/models/_m3gnet.py:301]: I will address this in next commit.

File: src/matgl/models/_m3gnet.py (hunk 2 above)
Comment (JiQi535) [reply src/matgl/models/_m3gnet.py:218]: If we return all layer outputs, the default behavior of forward method will be changed.
